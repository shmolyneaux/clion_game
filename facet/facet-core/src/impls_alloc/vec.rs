use crate::*;
use core::hash::Hash as _;

use alloc::boxed::Box;
use alloc::vec::Vec;

type VecIterator<'mem, T> = core::slice::Iter<'mem, T>;

unsafe impl<'a, T> Facet<'a> for Vec<T>
where
    T: Facet<'a>,
{
    const VTABLE: &'static ValueVTable = &const {
        ValueVTable::builder::<Self>()
            .type_name(|f, opts| {
                if let Some(opts) = opts.for_children() {
                    write!(f, "{}<", Self::SHAPE.type_identifier)?;
                    T::SHAPE.vtable.type_name()(f, opts)?;
                    write!(f, ">")
                } else {
                    write!(f, "{}<â‹¯>", Self::SHAPE.type_identifier)
                }
            })
            .default_in_place(|| Some(|target| unsafe { target.put(Self::default()) }))
            .clone_into(|| {
                if T::SHAPE.vtable.has_clone_into() {
                    Some(|src, dst| unsafe {
                        let mut new_vec = Vec::with_capacity(src.len());

                        let t_clone_into = <VTableView<T>>::of().clone_into().unwrap();

                        for item in src {
                            use crate::TypedPtrUninit;
                            use core::mem::MaybeUninit;

                            let mut new_item = MaybeUninit::<T>::uninit();
                            let uninit_item = TypedPtrUninit::new(new_item.as_mut_ptr());

                            (t_clone_into)(item, uninit_item);

                            new_vec.push(new_item.assume_init());
                        }

                        dst.put(new_vec)
                    })
                } else {
                    None
                }
            })
            .debug(|| {
                if T::SHAPE.vtable.has_debug() {
                    Some(|value, f| {
                        write!(f, "[")?;
                        for (i, item) in value.iter().enumerate() {
                            if i > 0 {
                                write!(f, ", ")?;
                            }
                            (<VTableView<T>>::of().debug().unwrap())(item, f)?;
                        }
                        write!(f, "]")
                    })
                } else {
                    None
                }
            })
            .partial_eq(|| {
                if T::SHAPE.vtable.has_partial_eq() {
                    Some(|a, b| {
                        if a.len() != b.len() {
                            return false;
                        }
                        for (item_a, item_b) in a.iter().zip(b.iter()) {
                            if !(<VTableView<T>>::of().partial_eq().unwrap())(item_a, item_b) {
                                return false;
                            }
                        }
                        true
                    })
                } else {
                    None
                }
            })
            .hash(|| {
                if T::SHAPE.vtable.has_hash() {
                    Some(|vec, hasher_this, hasher_write_fn| unsafe {
                        use crate::HasherProxy;
                        let t_hash = <VTableView<T>>::of().hash().unwrap_unchecked();
                        let mut hasher = HasherProxy::new(hasher_this, hasher_write_fn);
                        vec.len().hash(&mut hasher);
                        for item in vec {
                            (t_hash)(item, hasher_this, hasher_write_fn);
                        }
                    })
                } else {
                    None
                }
            })
            .marker_traits(|| {
                MarkerTraits::SEND
                    .union(MarkerTraits::SYNC)
                    .union(MarkerTraits::EQ)
                    .union(MarkerTraits::UNPIN)
                    .union(MarkerTraits::UNWIND_SAFE)
                    .union(MarkerTraits::REF_UNWIND_SAFE)
                    .intersection(T::SHAPE.vtable.marker_traits())
            })
            .build()
    };

    const SHAPE: &'static Shape<'static> = &const {
        Shape::builder_for_sized::<Self>()
            .type_identifier("Vec")
            .type_params(&[TypeParam {
                name: "T",
                shape: || T::SHAPE,
            }])
            .ty(Type::User(UserType::Opaque))
            .def(Def::List(
                ListDef::builder()
                    .vtable(
                        &const {
                            ListVTable::builder()
                                .init_in_place_with_capacity(|data, capacity| unsafe {
                                    data.put(Self::with_capacity(capacity))
                                })
                                .push(|ptr, item| unsafe {
                                    let vec = ptr.as_mut::<Self>();
                                    let item = item.read::<T>();
                                    (*vec).push(item);
                                })
                                .len(|ptr| unsafe {
                                    let vec = ptr.get::<Self>();
                                    vec.len()
                                })
                                .get(|ptr, index| unsafe {
                                    let vec = ptr.get::<Self>();
                                    let item = vec.get(index)?;
                                    Some(PtrConst::new(item))
                                })
                                .get_mut(|ptr, index| unsafe {
                                    let vec = ptr.as_mut::<Self>();
                                    let item = vec.get_mut(index)?;
                                    Some(PtrMut::new(item))
                                })
                                .as_ptr(|ptr| unsafe {
                                    let vec = ptr.get::<Self>();
                                    PtrConst::new(vec.as_ptr())
                                })
                                .as_mut_ptr(|ptr| unsafe {
                                    let vec = ptr.as_mut::<Self>();
                                    PtrMut::new(vec.as_mut_ptr())
                                })
                                .iter_vtable(
                                    IterVTable::builder()
                                        .init_with_value(|ptr| unsafe {
                                            let vec = ptr.get::<Self>();
                                            let iter: VecIterator<T> = vec.iter();
                                            let iter_state = Box::new(iter);
                                            PtrMut::new(Box::into_raw(iter_state) as *mut u8)
                                        })
                                        .next(|iter_ptr| unsafe {
                                            let state = iter_ptr.as_mut::<VecIterator<'_, T>>();
                                            state.next().map(|value| PtrConst::new(value))
                                        })
                                        .next_back(|iter_ptr| unsafe {
                                            let state = iter_ptr.as_mut::<VecIterator<'_, T>>();
                                            state.next_back().map(|value| PtrConst::new(value))
                                        })
                                        .dealloc(|iter_ptr| unsafe {
                                            drop(Box::from_raw(
                                                iter_ptr.as_ptr::<VecIterator<'_, T>>()
                                                    as *mut VecIterator<'_, T>,
                                            ));
                                        })
                                        .build(),
                                )
                                .build()
                        },
                    )
                    .t(|| T::SHAPE)
                    .build(),
            ))
            .build()
    };
}
