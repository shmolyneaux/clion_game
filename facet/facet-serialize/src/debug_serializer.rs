use crate::Serializer;

use alloc::vec::Vec;

struct DebugSerializer<W> {
    writer: W,
    need_comma: Vec<bool>,
}

#[derive(Debug)]
enum DebugError {
    Fmt(core::fmt::Error),
}

impl core::fmt::Display for DebugError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        core::fmt::Debug::fmt(self, f)
    }
}

impl core::error::Error for DebugError {}

impl From<core::fmt::Error> for DebugError {
    fn from(err: core::fmt::Error) -> Self {
        DebugError::Fmt(err)
    }
}

impl<'shape, W> Serializer<'shape> for DebugSerializer<W>
where
    W: core::fmt::Write,
{
    type Error = DebugError;

    fn serialize_u8(&mut self, value: u8) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_u16(&mut self, value: u16) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_u32(&mut self, value: u32) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_u64(&mut self, value: u64) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_u128(&mut self, value: u128) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_usize(&mut self, value: usize) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_i8(&mut self, value: i8) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_i16(&mut self, value: i16) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_i32(&mut self, value: i32) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_i64(&mut self, value: i64) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_i128(&mut self, value: i128) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_isize(&mut self, value: isize) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_f32(&mut self, value: f32) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_f64(&mut self, value: f64) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_bool(&mut self, value: bool) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{}", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_char(&mut self, value: char) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "\"{}\"", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_str(&mut self, value: &str) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "\"{}\"", value)?;
        self.set_comma();
        Ok(())
    }

    fn serialize_bytes(&mut self, value: &[u8]) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "[")?;

        self.need_comma.push(false);
        for byte in value.iter() {
            self.write_comma()?;
            write!(self.writer, "{}", byte)?;
            self.set_comma();
        }
        self.need_comma.pop();

        write!(self.writer, "]")?;
        self.set_comma();
        Ok(())
    }

    fn serialize_none(&mut self) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "null")?;
        self.set_comma();
        Ok(())
    }

    fn serialize_unit(&mut self) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "null")?;
        self.set_comma();
        Ok(())
    }

    fn serialize_unit_variant(
        &mut self,
        _variant_index: usize,
        variant_name: &str,
    ) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "\"{}\"", variant_name)?;
        self.set_comma();
        Ok(())
    }

    fn start_object(&mut self, _len: Option<usize>) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{{")?;
        self.need_comma.push(false);
        Ok(())
    }

    fn end_object(&mut self) -> Result<(), Self::Error> {
        self.need_comma.pop();
        write!(self.writer, "}}")?;
        self.set_comma();
        Ok(())
    }

    fn start_array(&mut self, _len: Option<usize>) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "[")?;
        self.need_comma.push(false);
        Ok(())
    }

    fn end_array(&mut self) -> Result<(), Self::Error> {
        self.need_comma.pop();
        write!(self.writer, "]")?;
        self.set_comma();
        Ok(())
    }

    fn start_map(&mut self, _len: Option<usize>) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "{{")?;
        self.need_comma.push(false);
        Ok(())
    }

    fn end_map(&mut self) -> Result<(), Self::Error> {
        self.need_comma.pop();
        write!(self.writer, "}}")?;
        self.set_comma();
        Ok(())
    }

    fn serialize_field_name(&mut self, name: &str) -> Result<(), Self::Error> {
        self.write_comma()?;
        write!(self.writer, "\"{}\":", name)?;
        if let Some(need_comma) = self.need_comma.last_mut() {
            *need_comma = false;
        }
        Ok(())
    }
}

impl<W> DebugSerializer<W>
where
    W: core::fmt::Write,
{
    fn write_comma(&mut self) -> Result<(), DebugError> {
        if let Some(&true) = self.need_comma.last() {
            write!(self.writer, ", ")?;
        }
        Ok(())
    }

    fn set_comma(&mut self) {
        if let Some(need_comma) = self.need_comma.last_mut() {
            *need_comma = true;
        }
    }
}

#[cfg(test)]
mod tests {
    use facet::Facet;
    use facet_reflect::Peek;

    use crate::serialize_iterative;

    use super::DebugSerializer;

    #[derive(Facet)]
    struct FooBarBaz {
        foo: u32,
        bar: String,
        baz: bool,
    }

    #[test]
    fn test_serialize_debug() -> eyre::Result<()> {
        facet_testhelpers::setup();

        let val = FooBarBaz {
            foo: 42,
            bar: "Hello".to_string(),
            baz: true,
        };
        let peek = Peek::new(&val);

        let mut s = String::new();
        let mut serializer = DebugSerializer {
            writer: &mut s,
            need_comma: vec![false],
        };
        serialize_iterative(peek, &mut serializer)?;
        #[cfg(not(miri))]
        insta::assert_snapshot!(s);

        Ok(())
    }

    #[derive(Facet)]
    struct FooBarBazContainer {
        foo_bar_baz: FooBarBaz,
        other_things: Vec<SomeEnum>,
        maybe_value: Option<u32>,
        definitely_none: Option<String>,
    }

    #[derive(Facet)]
    #[repr(u8)]
    #[allow(dead_code)]
    enum SomeEnum {
        Variant1,
        Variant2(u32, String),
        Variant3 { field1: u32, field2: String },
    }

    #[test]
    fn test_serialize_debug_container() {
        facet_testhelpers::setup();

        let val = FooBarBazContainer {
            foo_bar_baz: FooBarBaz {
                foo: 42,
                bar: "Hello".to_string(),
                baz: true,
            },
            other_things: vec![
                SomeEnum::Variant1,
                SomeEnum::Variant2(123, "TupleVariant".to_string()),
                SomeEnum::Variant3 {
                    field1: 456,
                    field2: "StructVariant".to_string(),
                },
            ],
            maybe_value: Some(42),
            definitely_none: None,
        };
        let peek = Peek::new(&val);

        let mut s = String::new();
        let mut serializer = DebugSerializer {
            writer: &mut s,
            need_comma: vec![false],
        };
        serialize_iterative(peek, &mut serializer).unwrap();
        #[cfg(not(miri))]
        insta::assert_snapshot!(s);
    }

    #[derive(Facet)]
    #[facet(transparent)]
    #[repr(transparent)]
    struct Wrapper(String);

    #[test]
    fn test_serialize_transparent() {
        facet_testhelpers::setup();

        let val = Wrapper("TransparentValue".to_string());
        let peek = Peek::new(&val);

        let mut s = String::new();
        let mut serializer = DebugSerializer {
            writer: &mut s,
            need_comma: vec![false],
        };
        serialize_iterative(peek, &mut serializer).unwrap();
        #[cfg(not(miri))]
        insta::assert_snapshot!(s);
    }
}
