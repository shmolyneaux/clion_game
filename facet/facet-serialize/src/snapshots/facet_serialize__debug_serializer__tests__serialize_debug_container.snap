---
source: facet-serialize/src/debug_serializer.rs
expression: s
---
{"foo_bar_baz":{"foo":42, "bar":"Hello", "baz":true}, "other_things":["Variant1", {"Variant2":[123, "TupleVariant"]}, {"Variant3":{"field1":456, "field2":"StructVariant"}}], "maybe_value":42, "definitely_none":null}
